### DISCUSSION TITLE
Controlling the viewport of a line graph

### DISCUSSION THREAD

**Discussion by michaelfominwahoo** [patrykandpatrick/vico#577] [Questions]
Hi, I'm having a play with 2.0.0-alpha.7 on an app that graphs real/recorded/live data (wrt time) on one line, over a second reference/target line.
I want to control the viewport (or the camera/eye) programmatically by specifying the viewport's width and scroll location in the x-value domain.  
The hope is at the start of the "data tracking session" the viewport will be Left=0min Width=10min, and the real-line will start to fill in from the left.
At 5min (real-line 50% across the chart) I'd like the viewport to "window along" so the user's location will remain centered and they'll see 5min of their real-history line and 5min of their target-future line.
I'm also keen to allow the user to swipe/scroll left and right at my previously specified viewport width. I figure that in this "free scroll mode" I simply won't set the viewport, and when the user wants to return to auto-scroll, I will return to periodically setting the viewport.
Apologies for the lengthy discussion, I hope it makes sense :) Thanks in advance.

**Comment by patrickmichalik** [comment]
Hello! [Vico 2.0.0 Alpha 9](https://github.com/patrykandpatrick/vico/releases/tag/v2.0.0-alpha.9) should enable you to implement this. For the purposes of this comment, I’ll assume that one _x_ unit corresponds to one minute. To show 10 minutes’ worth of data, set `initialZoom` to `Zoom.x(10f)`, and `zoomEnabled` to false. To ensure that the maximum _x_ value is at least 10, use `AxisValueOverrider`. To have the viewport move as described, set `autoScroll` to `Scroll.Absolute.End`, and for `autoScrollCondition`, use an `AutoScrollCondition` implementation that’s based on `AutoScrollCondition.OnModelSizeIncreased` but has an additional check that determines if auto-scroll mode is on. The `scroll` and `animateScroll` functions of `VicoScrollState` and `ScrollHandler` facilitate programmatic scrolling, but please note that automatic scrolling should be used for scrolls performed based on `CartesianChartModel` changes. To keep track of the current scroll value and the maximum scroll value, use either `VicoScrollState#value` and `VicoScrollState#maxValue` or `ScrollHandler#addListener`. If you have any questions, feel free to ask, and please let me know if you’re using Vico with Jetpack Compose or the view system.
